common4j/src/main/com/microsoft/identity/common/java/logging/Logger.java

```diff
@@ -0,0 +1,446 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.logging;
+
+import com.microsoft.identity.common.java.util.StringUtil;
+import com.microsoft.identity.common.java.util.ThrowableUtil;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TimeZone;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import lombok.Synchronized;
+import lombok.experimental.Accessors;
+
+public class Logger {
+
+    private static final ExecutorService sLogExecutor = Executors.newSingleThreadExecutor();
+    private static final String DATE_FORMAT = "yyyy-MM-dd HH:mm:ss";
+
+    // Turn on the VERBOSE level logging by default.
+    @Setter()
+    @Getter()
+    @Accessors(prefix = "s")
+    private static LogLevel sLogLevel = LogLevel.VERBOSE;
+
+    // Disable to log PII by default.
+    @Setter()
+    @Getter()
+    @Accessors(prefix = "s")
+    private static boolean sAllowPii = false;
+
+    @Setter(onMethod_ = {@Synchronized})
+    @Accessors(prefix = "s")
+    private static String sPlatformString = "";
+
+    private static final ReentrantReadWriteLock sLoggersLock = new ReentrantReadWriteLock();
+
+    private static final Map<String, ILoggerCallback> sLoggers = new HashMap<>();
+
+    /**
+     * Enum class for LogLevel that the sdk recognizes.
+     */
+    public enum LogLevel {
+        /**
+         * Error level logging.
+         */
+        ERROR,
+        /**
+         * Warn level logging.
+         */
+        WARN,
+        /**
+         * Info level logging.
+         */
+        INFO,
+        /**
+         * Verbose level logging.
+         */
+        VERBOSE,
+        /**
+         * Undefined. Should be used in test only.
+         */
+        UNDEFINED,
+    }
+
+    // Visible for testing
+    static synchronized void resetLogger() {
+        sLoggersLock.writeLock().lock();
+        try {
+            sLoggers.clear();
+            sAllowPii = false;
+            sPlatformString = "";
+            sLogLevel = LogLevel.VERBOSE;
+        } finally {
+            sLoggersLock.writeLock().unlock();
+        }
+    }
+
+    public static void setLogger(@NonNull String identifier,
+                                 ILoggerCallback callback) {
+        sLoggersLock.writeLock().lock();
+        try {
+            if (callback == null &&
+                    sLoggers.containsKey(identifier)) {
```

Comment:
Do we need to `constainsKey`? I don't think so

Reply:
Yes. This means "if this identifier is already in the map, remove the old one *first*"

Reply:
ah.. I got what you're trying to say. nvm.